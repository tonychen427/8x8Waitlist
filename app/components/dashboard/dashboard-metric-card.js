/* @flow */

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import {
  StyleSheet,
  Text,
  View,
} from 'react-native'
import { intlShape, injectIntl } from 'react-intl'
import Icon from 'react-native-vector-icons/SimpleLineIcons'
import * as colors from './utils/colors'

const styles = StyleSheet.create({
  container: {
    flex: 1,
    shadowColor: colors.black,
    shadowOffset: {
      height: 1,
      width: 0,
    },
    shadowOpacity: 0.2,
    shadowRadius: 2,
    backgroundColor: colors.white,
    borderRadius: 6,
    padding: 16,
    margin: 8,
    display: 'flex',
    flexDirection: 'column',
  },
  header: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
  },
  title: {
    fontWeight: 'bold',
    paddingLeft: 8,
  },
  content: {
    paddingTop: 16,
  },
  row: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  values: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-end',
  },
  trend: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-end',
  },
})

export const DashboardMetricCard = props => (
  <View style={styles.container}>
    <View style={styles.header}>
      <View>
        <Icon
          name={props.iconName}
          color={colors.darkGrey}
          size={20}
        />
      </View>
      <Text style={styles.title}>{props.title}</Text>
    </View>
    <View style={styles.content}>
      <View style={styles.row}>
        <View><Text>{'Today'}</Text></View>
        <View style={styles.values}>
          <View>
            <Text>{renderMoney(props.intl, props.todayValue, 'EUR')}</Text>
          </View>
          {props.showTrend
            ? renderTrend(props.intl, props.yesterdayValue, props.todayValue)
            : null
          }
        </View>
      </View>
      <View style={styles.row}>
        <View><Text>{'This week'}</Text></View>
        <View style={styles.values}>
          <View>
            <Text>{renderMoney(props.intl, props.weekValue, 'EUR')}</Text>
          </View>
          {props.showTrend
            ? renderTrend(props.intl, props.lastWeekValue, props.weekValue)
            : null
          }
        </View>
      </View>
      <View style={styles.row}>
        <View><Text>{'This month'}</Text></View>
        <View style={styles.values}>
          <View>
            <Text>{renderMoney(props.intl, props.monthValue, 'EUR')}</Text>
          </View>
          {props.showTrend
            ? renderTrend(props.intl, props.lastMonthValue, props.monthValue)
            : null
          }
        </View>
      </View>
    </View>
  </View>
)

DashboardMetricCard.displayName = 'DashboardMetricCard'
DashboardMetricCard.propTypes = {
  title: PropTypes.string.isRequired,
  iconName: PropTypes.string.isRequired,
  todayValue: PropTypes.number.isRequired,
  weekValue: PropTypes.number.isRequired,
  monthValue: PropTypes.number.isRequired,
  yesterdayValue: PropTypes.number.isRequired,
  lastWeekValue: PropTypes.number.isRequired,
  lastMonthValue: PropTypes.number.isRequired,
  showTrend: PropTypes.bool.isRequired,

  // connected
  intl: intlShape.isRequired,
}

export default injectIntl(DashboardMetricCard)


function renderMoney (intl, value, currencyCode) {
  return intl.formatNumber(
    value / 100,
    {
      style: 'currency',
      currency: currencyCode,
      currencyDisplay: 'symbol',
    }
  )
}

function renderTrend (intl, prevValue, nextValue) {
  if (prevValue === 0)
    return (
      <Text>{'N/A'}</Text>
    )

  const indicatorNumber = calculateIndicatorNumber(prevValue, nextValue)

  let valueColor
  let valueIcon
  if (indicatorNumber < 0) {
    valueColor = 'red'
    valueIcon = 'arrow-down'
  }
  else if (indicatorNumber >= 0) {
    valueColor = 'green'
    valueIcon = 'arrow-up'
  }

  if (!valueIcon)
    return (
      <Text>{'N/A'}</Text>
    )

  return (
    <View style={styles.trend}>
      <Icon
        name={valueIcon}
        color={colors[valueColor]}
        size={20}
      />
      <Text>{`${indicatorNumber}%`}</Text>
    </View>
  )
}

function calculateIndicatorNumber (oldVal, newVal) {
  const diffPercentage = ((newVal - oldVal) / oldVal) * 100

  // Check if the calculated value has decimal places.
  // If so returns the value fixed to 2 decimals.
  if (Math.round(diffPercentage) === diffPercentage)
    return diffPercentage

  return diffPercentage.toFixed(2)
}
