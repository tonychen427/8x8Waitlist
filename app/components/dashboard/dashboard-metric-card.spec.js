import { Text } from 'react-native'
// Note: test renderer must be required after react-native.
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from 'react-test-renderer'
import React from 'react'
import { DashboardMetricCard } from './dashboard-metric-card'

const createTestProps = props => ({
  title: 'The title',
  iconName: 'chart',
  todayValue: 10,
  weekValue: 20,
  monthValue: 30,
  yesterdayValue: 5,
  lastWeekValue: 10,
  lastMonthValue: 15,
  showTrend: true,
  intl: {
    formatNumber: amount => String(amount),
  },
  ...props,
})

describe('rendering', () => {
  let wrapper
  beforeEach(() => {
    const props = createTestProps()
    wrapper = renderer.create(
      <DashboardMetricCard {...props} />,
    ).toJSON()
  })
  describe('with trend', () => {
    it('should render values with calculated trend', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('without trend', () => {
    beforeEach(() => {
      const props = createTestProps({ showTrend: false })
      wrapper = renderer.create(
        <DashboardMetricCard {...props} />,
      ).toJSON()
    })
    it('should render only values', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })
})
