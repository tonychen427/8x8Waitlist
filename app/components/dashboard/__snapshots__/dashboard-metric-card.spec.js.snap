// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering with trend should render values with calculated trend 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderRadius": 6,
      "display": "flex",
      "flex": 1,
      "flexDirection": "column",
      "margin": 8,
      "padding": 16,
      "shadowColor": undefined,
      "shadowOffset": Object {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.2,
      "shadowRadius": 2,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View>
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#999999",
              "fontSize": 20,
            },
            undefined,
            Object {
              "fontFamily": "simple-line-icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        
      </Text>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontWeight": "bold",
          "paddingLeft": 8,
        }
      }
    >
      The title
    </Text>
  </View>
  <View
    style={
      Object {
        "paddingTop": 16,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          Today
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "flex-end",
          }
        }
      >
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            0.1
          </Text>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "flex-end",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#20AD92",
                  "fontSize": 20,
                },
                undefined,
                Object {
                  "fontFamily": "simple-line-icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
              ]
            }
          >
            
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            100%
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          This week
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "flex-end",
          }
        }
      >
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            0.2
          </Text>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "flex-end",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#20AD92",
                  "fontSize": 20,
                },
                undefined,
                Object {
                  "fontFamily": "simple-line-icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
              ]
            }
          >
            
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            100%
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          This month
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "flex-end",
          }
        }
      >
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            0.3
          </Text>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "flex-end",
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": "#20AD92",
                  "fontSize": 20,
                },
                undefined,
                Object {
                  "fontFamily": "simple-line-icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
              ]
            }
          >
            
          </Text>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            100%
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`rendering without trend should render only values 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderRadius": 6,
      "display": "flex",
      "flex": 1,
      "flexDirection": "column",
      "margin": 8,
      "padding": 16,
      "shadowColor": undefined,
      "shadowOffset": Object {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.2,
      "shadowRadius": 2,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View>
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#999999",
              "fontSize": 20,
            },
            undefined,
            Object {
              "fontFamily": "simple-line-icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        
      </Text>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontWeight": "bold",
          "paddingLeft": 8,
        }
      }
    >
      The title
    </Text>
  </View>
  <View
    style={
      Object {
        "paddingTop": 16,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          Today
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "flex-end",
          }
        }
      >
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            0.1
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          This week
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "flex-end",
          }
        }
      >
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            0.2
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          This month
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "flex-end",
          }
        }
      >
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
          >
            0.3
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
